kyo-core/shared/src/main/scala/kyo/layers.scala

```diff
@@ -0,0 +1,144 @@
+package kyo
+
+object layers {
+
+  trait Layer[Sin, Sout] { self =>
+    def run[T, S](effect: T > (S with Sout))(implicit fl: Flat[T]): T > (S with Sin)
+
+    final def ++[Sin1, Sout1](other: Layer[Sin1, Sout1]): Layer[Sin with Sin1, Sout with Sout1] =
+      new Layer[Sin with Sin1, Sout with Sout1] {
+        override def run[T, S](
+            effect: T > (S with Sout with Sout1)
+        )(
+            implicit fl: Flat[T]
+        ): T > (S with Sin with Sin1) = {
+          val selfRun: T > (S with Sout1 with Sin) =
+            self.run[T, S with Sout1](effect: T > (S with Sout1 with Sout))
+          val otherRun: T > (S with Sin with Sin1) = other.run[T, S with Sin](selfRun)
+          otherRun
+        }
+      }
+
+    final def >>>[Sin1, Sout1](other: Layer[Sin1, Sout1])(implicit
+        ap: ApplyLayer[Sout, Sin1]
```

Comment:
Would it be possible to express the type-level transformations in the method signature itself instead of relying on the implicit mechanism? I imagine it's not the case, just to double check.

Reply:
I'm not sure. Basically some mechanism is needed to extract the intersection between the effects that one layer handles and the effects that another layer needs to be handled, and keep track of the remaining effects of each. I'm pretty sure I can least simplify and clarify `ApplyLayer`, but I can't think of a way to produce the same functionality with just type bounds. I'll give it some more thought though.

Reply:
sounds good! the current version is ok with me btw

Reply:
I wasn't able to figure out an alternative, but I did improve the type class used to chain layers together. Now it's called `ChainLayer`, and it's easier to follow because it reverses the ZIO approach. We can now think of it in this way: the output effects of the first layer are the input effects of the second layer. The main function of `ChainLayer`, then, is to derive which outputs effects from the first layer remain unhandled.